src/codeql.ts

```diff
@@ -76,51 +77,97 @@ export async function runQuery(
-  const compatibleQueryKinds = bqrsInfo.compatibleQueryKinds;
-  const queryMetadata = await getQueryMetadata(
+  // Calculate query run information like BQRS file paths, etc.
+  const queryPackRunResults = await getQueryPackRunResults(
     codeql,
-    await getRemoteQueryPackDefaultQuery(codeql, queryPack),
+    databaseName,
+    queryPaths,
+    queryPack,
+    queryPackName,
   );
 
   const sourceLocationPrefix = await getSourceLocationPrefix(codeql);
-  const sarifOutputType = getSarifOutputType(
-    queryMetadata,
-    compatibleQueryKinds,
+
+  const shouldGenerateSarif = await queryPackSupportsSarif(
+    codeql,
+    queryPackRunResults,
   );
+
   let resultCount: number;
   let sarifFilePath: string | undefined;
-  if (sarifOutputType !== undefined) {
+  if (shouldGenerateSarif) {
     const sarif = await generateSarif(
       codeql,
-      bqrsFilePath,
       nwo,
-      queryMetadata,
-      sarifOutputType,
       databaseName,
-      sourceLocationPrefix,
+      queryPackName,
       databaseSHA,
     );
     resultCount = getSarifResultCount(sarif);
     sarifFilePath = path.join("results", "results.sarif");
     fs.writeFileSync(sarifFilePath, JSON.stringify(sarif));
   } else {
-    resultCount = getBqrsResultCount(bqrsInfo);
+    resultCount = queryPackRunResults.totalResultsCount;
   }
 
+  const bqrsFilePaths = await adjustBqrsFiles(queryPackRunResults);
+
   return {
     resultCount,
     databaseSHA,
+    databaseName,
     sourceLocationPrefix,
-    bqrsFilePath,
+    bqrsFilePaths,
     sarifFilePath,
   };
 }
 
+async function adjustBqrsFiles(
+  queryPackRunResults: QueryPackRunResults,
+): Promise<string[]> {
+  if (queryPackRunResults.queries.length === 1) {
+    // If we have a single query, move the BQRS file to "results.bqrs" in order to
+    // maintain backwards compatibility with the VS Code extension, since it expects
+    // the BQRS file to be at the top level and be called "results.bqrs".
+    const currentBqrsFilePath = path.join(
+      queryPackRunResults.resultsBasePath,
+      queryPackRunResults.queries[0].relativeBqrsFilePath,
+    );
+    const newBqrsFilePath = path.join("results", "results.bqrs");
+    await fs.promises.rename(currentBqrsFilePath, newBqrsFilePath);
+    return [newBqrsFilePath];
+  }
+
+  // If we have multiple queries, move the BQRS files to the correct location and
+  // return the new paths.
+  return await Promise.all(
+    queryPackRunResults.queries.map(async (query) => {
+      const newPath = await moveBqrsFileToResultsDir(
+        queryPackRunResults.resultsBasePath,
+        query.relativeBqrsFilePath,
+      );
+      return newPath;
+    }),
+  );
+}
+
+async function moveBqrsFileToResultsDir(
+  resultsBasePath: string,
+  relativeBqrsFilePath: string,
+): Promise<string> {
+  const oldPath = path.join(resultsBasePath, relativeBqrsFilePath);
+  const newPath = path.join("results", relativeBqrsFilePath);
+
+  if (!fs.existsSync(path.dirname(newPath))) {
```

Comment:
This could have a high potential for a concurrency bug where the directory is created after we check for it. We'll be performing this operation for many files at once in parallel.

Do we gain much performance by doing all the calls to `moveBqrsFileToResultsDir` at once or would it be safer to make each call synchronous, i.e. use a for loop instead of `Promise.all`?

Reply:
I think you can just call `fs.promises.mkdir` unconditionally since the `recursive` option is set to `true`. According to [the documentation](https://nodejs.org/docs/latest-v20.x/api/fs.html#fspromisesmkdirpath-options) "calling `fsPromises.mkdir()` when `path` is a directory that exists results in a rejection only when `recursive` is `false`."